File: code/code.go
package code

import (
	"archive/tar"
	"compress/gzip"
	"encoding/base64"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/exec"

	"github.com/aws/aws-lambda-go/events"
)

// CodeResponse contains the response from
// executing the Go code
type CodeResponse struct {
	ExitCode string `json:"exit_code"`
	Output   string `json:"output"`
}

func ExtractTarGz(gzipStream io.Reader) {
	uncompressedStream, err := gzip.NewReader(gzipStream)
	if err != nil {
		log.Fatal("ExtractTarGz: NewReader failed")
	}

	tarReader := tar.NewReader(uncompressedStream)

	for {
		header, err := tarReader.Next()

		if err == io.EOF {
			break
		}

		if err != nil {
			log.Fatalf("ExtractTarGz: Next() failed: %s", err.Error())
		}

		switch header.Typeflag {
		case tar.TypeDir:
			if err := os.Mkdir(header.Name, 0755); err != nil {
				log.Fatalf("ExtractTarGz: Mkdir() failed: %s", err.Error())
			}
		case tar.TypeReg:
			outFile, err := os.Create(header.Name)
			if err != nil {
				log.Fatalf("ExtractTarGz: Create() failed: %s", err.Error())
			}
			if _, err := io.Copy(outFile, tarReader); err != nil {
				log.Fatalf("ExtractTarGz: Copy() failed: %s", err.Error())
			}
			outFile.Close()

		default:
			log.Fatalf("ExtractTarGz: uknown type")
		}

	}
}

func setupGoEnvironment() error {
	path := os.Getenv("PATH")
	os.Setenv("PATH", path+":/tmp/go/bin")
	os.Setenv("GOROOT", "/tmp/go")
	os.Setenv("GOPATH", "/tmp")
	os.Setenv("GOCACHE", "/tmp/go-cache")

	r, err := os.Open("./code/go.tar.gz")
	if err != nil {
		fmt.Println("error")
		return err
	}
	ExtractTarGz(r)

	return nil
}

// ExecuteCode does the job of taking the Go code that has
// been sent to API from a snippet and executing it before
// returning the response
func ExecuteCode(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {

	fmt.Println("Received body: ", request.Body)
	body, _ := base64.StdEncoding.DecodeString(request.Body)
	fmt.Println(string(body))

	err := setupGoEnvironment()
	if err != nil {
		log.Fatalf("Setting up Go Env Failed")
		return events.APIGatewayProxyResponse{
			Body:       "Failed to setup Go Environment",
			Headers:    map[string]string{"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"},
			StatusCode: 503,
		}, nil
	}

	out, err := exec.Command("go", "version").CombinedOutput()
	if err != nil {
		fmt.Println(err.Error())

		log.Fatalf("executing go version failed %s\n", err)
		return events.APIGatewayProxyResponse{
			Body:       "Failed to run go version",
			Headers:    map[string]string{"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"},
			StatusCode: 503,
		}, nil
	}
	fmt.Println(string(out))

	// the WriteFile method returns an error if unsuccessful
	err = ioutil.WriteFile("main.go", body, 0777)
	// handle this error
	if err != nil {
		// print it out
		fmt.Println(err)
		return events.APIGatewayProxyResponse{
			Body:       "Failed to write main.go",
			Headers:    map[string]string{"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"},
			StatusCode: 503,
		}, nil
	}

	out, err = exec.Command("go", "run", "main.go").CombinedOutput()
	if err != nil {
		fmt.Println(err.Error())
		return events.APIGatewayProxyResponse{
			Body:       "Failed to run main.go",
			Headers:    map[string]string{"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"},
			StatusCode: 503,
		}, nil
	}

	fmt.Println(string(out))

	return events.APIGatewayProxyResponse{
		Body:       string(out),
		Headers:    map[string]string{"Content-Type": "application/json", "Access-Control-Allow-Origin": "*"},
		StatusCode: 200,
	}, nil
}

File: code/code.py
import subprocess
import json
import tempfile
import sys
import os
import tarfile

def create_temp_file(event):
    tempFile = tempfile.NamedTemporaryFile(delete=False, dir="/tmp", suffix=".go")
    with tempFile as fp:
        tempFile.write(bytes(event["body"], 'utf-8'))

    return tempFile

def run_go_code(temp_file):
    print(os.listdir())
    try:
        if not os.path.exists("/tmp/go"):
            go_code = tarfile.open("./code/go.tar.gz", "r:gz")
            print(go_code.getmembers())
            print(go_code.getnames())
            go_code.extractall("/tmp")
            go_code.close()
        print(os.listdir("/tmp/go"))
    except:
        e = sys.exc_info()[0]
        print(e)

    my_env = os.environ.copy()
    my_env["PATH"] = "/usr/sbin:/sbin:/tmp/go/bin"
    my_env["GOROOT"] = "/tmp/go"
    my_env["GOPATH"] = "/tmp"

    try:
        args = ["tar", "-C", "/tmp/go2", "./code/go.tar.gz"]
        popen = subprocess.Popen(args, stdout=subprocess.PIPE, env=my_env)
        popen.wait()
        print(os.listdir("/tmp/go2"))
    except:
        e = sys.exc_info()[0]
        print(e)

    args = ["go", "version"]
    popen = subprocess.Popen(args, stdout=subprocess.PIPE, env=my_env)
    popen.wait()
        
    args = ["go", "run", temp_file.name]
    popen = subprocess.Popen(args, stdout=subprocess.PIPE, env=my_env)
    popen.wait()


def lambda_handler(event, context):
    temp_file = create_temp_file(event)
    run_go_code(temp_file)

    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!')
    }

# def main():
#     event = {}
#     event["body"] = """package main

#     import "fmt"

#     func main() {

#         mymap := make(map[string]int)

#         mymap["elliot"] = 25

#         // we can use this if statement to check to see if 
#         // a given key "elliot" exists within a map in Go
#         if _, ok := mymap["elliot"]; ok {
#             // the key 'elliot' exists within the map
#             fmt.Println(mymap["elliot"])
#         }
#     }""" 

#     status = lambda_handler(event, None)
#     print(status)

# if __name__ == '__main__':
#     main()